src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -331,9 +368,67 @@ public object GetOrRegisterObjectForComInstance(IntPtr externalComObject, Create
             if (flags.HasFlag(CreateObjectFlags.Aggregation))
                 throw new NotImplementedException();
 
-            object? wrapperMaybeLocal = wrapperMaybe;
-            retValue = null;
-            throw new NotImplementedException();
+            if (flags.HasFlag(CreateObjectFlags.Unwrap))
+            {
+                var comInterfaceDispatch = impl.TryGetComInterfaceDispatch(externalComObject);
+                if (comInterfaceDispatch != null)
+                {
+                    retValue = ComInterfaceDispatch.GetInstance<object>(comInterfaceDispatch);
+                    return true;
+                }
+            }
+
+            if (!flags.HasFlag(CreateObjectFlags.UniqueInstance))
+            {
+                using (LockHolder.Hold(impl._lock))
+                {
+                    if (impl._rcwCache.TryGetValue(externalComObject, out var handle))
+                    {
+                        if (handle.IsAllocated)
+                        {
+                            retValue = handle.Target;
+                            return false;
+                        }
+                    }
+                }
+            }
+
+            retValue = impl.CreateObject(externalComObject, flags);
+            if (retValue == null)
+            {
+                // If ComWrappers instance cannot create wrapper, we can do nothing here.
+                return false;
+            }
+
+            if (flags.HasFlag(CreateObjectFlags.UniqueInstance))
+            {
+                // No need to cache NativeObjectWrapper for unique instances. They are not cached.
+                return true;
+            }
+
+            using (LockHolder.Hold(impl._lock))
+            {
+                if (impl._rcwCache.TryGetValue(externalComObject, out var existingHandle))
+                {
+                    if (existingHandle.IsAllocated)
+                    {
+                        retValue = existingHandle.Target;
+                    }
+                }
+                else
+                {
+                    GCHandle proxyHandle = GCHandle.Alloc(retValue, GCHandleType.Weak);
+                    GCHandle comWrappersHandle = GCHandle.Alloc(impl, GCHandleType.Weak);
```

Comment:
Why do you need the ComWrappers reference to be weak?

Reply:
Yes. If ComWrappers garbange collected, my understanding that `NativeObjectWrapper` instances in the `ConditionalWeakTable` would not be immediately collected and just scheduled to run finalizers (?) which may happens a bit after or maybe in different order. So I should not remove anything from `_rcwCache` in that case becasue object already dead.

Reply:
I do not see the problem. If you change `NativeObjectWrapper` to have a regular reference to the owning ComWrappers instance,, what is not going to work? Is there a test that is going to fail?

Reply:
As always you are right. I was over-complicating stuff. Remove handle on `ComWrappers`